### PR TITLE
BUG: styler multiindex hiding indexes and columns alignment

### LINKED ISSUE #43637: BUG: `styler.to_latex` does hide rows after `hide_index([subset])`
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
>>> df = DataFrame([[1], [2]])
>>> print(df.style.hide_index([0]).to_latex())
\begin{tabular}{lr}
 & 0 \\
 \\
1 & 2 \\
\end{tabular}
```

### Issue Description

its the header cells that render as empty spaces creating an unwanted line. 

### Expected Behavior

no unwanted line.

### Installed Versions

master

### LINKED ISSUE #43644: BUG: styler hide_index hide_columns misaligns multiindex
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
midx = MultiIndex.from_product([["A", "B"], ["a", "b"], ["X", "Y"]], names=["zero", "one", "two"])
cidx = MultiIndex.from_product([["a", "b"], ["C", "D"], ["V", "W"]], names=["zero", "one", "two"])
df = DataFrame(np.arange(64).reshape(8,8), index=midx, columns=cidx)
styler = Styler(df)
styler.hide_index(level=1).hide_columns(level=0)
styler.hide_index([("A", "a", "X"), ("A", "b", "X")])
styler.hide_columns([("a", "C", "W"), ("a", "D", "V")])
styler
```

### Issue Description

When using a MultiIndex and combining with hiding different levels and hiding specific rows/columns, some bugs appear.

Output is quite strange on second row for HTML output.

Screen Shot 2021-09-18 at 19 40 34

Also using LaTeX output:  
```
>>> styler.to_latex()
\begin{tabular}{llrrrrrr}
 & one & C & D & \multicolumn{2}{r}{C} & \multicolumn{2}{r}{D} \\
 & two & V & W & V & W & V & W \\
zero & two &  &  &  &  &  &  \\
\multirow[c]{2}{*}{A} &  & Y & 8 & 11 & 12 & 13 & 14 & 15 \\
 &  & Y & 24 & 27 & 28 & 29 & 30 & 31 \\
\multirow[c]{4}{*}{B} & \multirow[c]{2}{*}{} & X & 32 & 35 & 36 & 37 & 38 & 39 \\
 &  & Y & 40 & 43 & 44 & 45 & 46 & 47 \\
 & \multirow[c]{2}{*}{} & X & 48 & 51 & 52 & 53 & 54 & 55 \\
 &  & Y & 56 & 59 & 60 & 61 & 62 & 63 \\
\end{tabular}
```

The above is a LaTeX error but if the number of columns is expanded to 9 then it renders like:
Screen Shot 2021-09-18 at 19 43 51

### Expected Behavior

Clearly we want to align all the datapoints and have the hidden levels correctly hidden.

### Installed Versions

master

### LINKED ISSUE #43703: BUG: styler render trimming rows does not work with `hide_index`
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
pd.options.styler.render.max_rows = 2
df = DataFrame(np.random.randn(5,1))
df.style.hide_index(df.index[:2])
```

### Issue Description

This does not render any rows since it loops through the first 2 rows, which are hidden, but then doesn't progress the loop passed index 2.

### Expected Behavior

Hidden rows should be deducted from the row count. The dataframe should show rows 3 and 4 above.

### Installed Versions

master

### PR DESCRIPTION
**Pull request by attack68**
#### Follows #43686

- [x] closes #43637 
- [x] closes #43644
- [x] closes #43703
- [x] tests added / passed
- [x] whatsnew entry (note part of the bug seems to have been created by hiding `levels` which is a new feature in 1.4.0).

This PR has two parts:

1) Addressing the bug fixes as per above.

The key change made here was to add a single line: `if r not in self.hidden_rows:` in the loop, and in `_translate_latex` update for the new object structure that is input to it.

2) Refactoring the code so that the main part of the code is now easier to grok:

```
def _translate(...):
    # calls _translate_header()
    # callls _translate_body()

def _translate_header(...):
   # calls _generate_col_header_row()
   # calls _generate_index_names_row()

def _translate_body(...):
   # calls _generate_trimmed_row()
   # calls _generate_body_row()
```

And tests were update and comprehensive tests for the issues added.

### CODE REVIEW

File: pandas/io/formats/style_render.py [hunk 1]
```diff
@@ -497,119 +497,131 @@ def _translate_body(
...
-                    _element(
-                        "td",
-                        f"{data_class} col{c} {trimmed_row_class}",
-                        "...",
-                        (c not in self.hidden_columns),
-                        attributes="",
-                    )
-                    for c in range(max_cols)
-                ]
+            if r not in self.hidden_rows:
```
Comment (attack68) [inline pandas/io/formats/style_render.py:500]: only added this line to this method and all code below it was indended one tab.

File: pandas/io/formats/style_render.py [hunk 2]
```diff
@@ -619,8 +631,16 @@ def _translate_latex(self, d: dict) -> None:
             ]
             for r, row in enumerate(d["head"])
         ]
+
         body = []
-        for r, row in enumerate(d["body"]):
+        index_levels = self.data.index.nlevels
+        for r, row in zip(
+            [r for r in range(len(self.data.index)) if r not in self.hidden_rows],
```
Comment (jreback) [inline pandas/io/formats/style_render.py:638]: might be worth a property to have this

File: pandas/io/formats/style_render.py (hunk 1 above)
Comment (jreback) [reply pandas/io/formats/style_render.py:500]: might be worth splitting up translate_body to some function calls to make it simpler to grok (future)

Comment (attack68) [reply pandas/io/formats/style_render.py:657]: i will finish making this clean up after #43686

**Comment by attack68** [comment]
because of moving the hiding functions to outer loop:

```
       before           after         ratio
     [556f4379]       [c22cf0d8]
     <master>         <bug_styler_multiindex_hiding>
-      6.47±0.2ms       5.34±0.2ms     0.83  io.style.Render.time_apply_format_hide_render(12, 12)
-      11.0±0.1ms       8.41±0.1ms     0.77  io.style.Render.time_apply_format_hide_render(24, 12)
-      45.2±0.7ms         26.2±8ms     0.58  io.style.Render.time_apply_format_hide_render(24, 120)
-        64.2±2ms       29.0±0.9ms     0.45  io.style.Render.time_apply_format_hide_render(36, 120)
-      25.7±0.4ms      11.1±0.09ms     0.43  io.style.Render.time_apply_format_hide_render(12, 120)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -143,11 +143,16 @@ class Styler(StylerRenderer):
         uses ``pandas.options.styler.format.formatter``.
 
         .. versionadded:: 1.4.0
+    css : dict, optional
+        A dict of strings used to replace the default CSS classes described below.
```
Comment (jreback) [inline pandas/io/formats/style.py:147]: needs versionadded

File: pandas/io/formats/style.py [hunk 4]
```diff
@@ -143,11 +143,16 @@ class Styler(StylerRenderer):
...
 
     Attributes
     ----------
     env : Jinja2 jinja2.Environment
-    template : Jinja2 Template
+    template_html : Jinja2 Template
```
Comment (jreback) [inline pandas/io/formats/style.py:152]: useful to document these?

**Review by jreback — CHANGES_REQUESTED** [review]
prob worth having a set_css. how does this interact with set_table_styles? can we only have 1.

File: pandas/io/formats/style.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style.py:152]: there is something in the user guide. can do as follow on if you think needs more.

File: pandas/io/formats/style.py (hunk 3 above)
Comment (attack68) [reply pandas/io/formats/style.py:147]: this PR follows #43686, so effectively splits this up into manageable chunks. Can you review that one first.

**Review by jreback — APPROVED** [review]

**Pull request merged**